osbuild/sources.py

```diff
@@ -51,17 +54,66 @@ class SourceService(host.Service):
     """Source host service"""
 
     @abc.abstractmethod
-    def download(self, items, cache, options):
-        pass
+    def fetch_one(self, checksum, desc) -> None:
+        """Performs the actual fetch of an element described by its checksum and its descriptor"""
+
+    def config_download(self) -> None:
+        """Hook to set config for the download affecting all the elements, such as parallelism or special caches."""
+
+    def exists(self, checksum, _desc) -> bool:
+        """Returns True if the item to download is in cache. """
+        return os.path.isfile(f"{self.cache}/{checksum}")
+
+    # pylint: disable=[no-self-use]
+    def transform(self, checksum, desc) -> Tuple:
+        """Modify the input data before downloading. By default only transforms an item object to a Tupple."""
+        return checksum, desc
+
+    def _unpack(self, items: List[Tuple]) -> (List, List):
+        """Change a list of tuples [(hash:desc), ...] to two lists of the same sizes, [hash, ...], [desc, ...]."""
+        return map(list, zip(*items))
+
+    def _prepare(self, items) -> List[Tuple]:
+        """Prepare every item transforming them before downloading."""
+        return [self.transform(checksum, desc) for checksum, desc in items.items()]
+
+    def _filter(self, items: Dict) -> Dict:
+        """Only keep the items that aren't alredy in the cache."""
+        return {checksum: desc for checksum, desc in items.items() if not self.exists(checksum, desc)}
```

Comment:
Nice! But I think we can shorten this and move them into `download` because they are basically just `map` and `filter` and only used in one place; we can use a little helper method to do the unpacking.

```python3
def apply_unpack(f: Callable) -> Callable:
    def inner(t: tuple):
       return f(t[0], t[1])
     return inner

items = filter(apply_unpacked(self.exists), items)  # filter existing
items = map(apply_unpacked(self.transform, items) # apply item preparations 

with concurrent.futures.ThreadPoolExecutor(max_workers=self.num_parallel_fetch) as executor:
             for _ in executor.map(self.fetch_one, *list(items)):
```

[To properly type the `Callable` we would have to use `T = typingTypeVar('T')` and:
```python3
T = TypeVar('T')
R = TypeVar('R')

def apply_unpack(f: Callable[str, T, R]) -> Callable[tuple[str, T, R]]:
    def inner(t: tuple[str, T]) -> Callable[str, T, R]:
       return f(t[0], t[1])
     return inner
```
But that seems excessive.